docs/adr/0006_component_constructors.md

```diff
@@ -0,0 +1,334 @@
+      If an access method is defined, it is applied directly to the component descriptor candidate.
+
+    - **"By Value" Resources**  
+      If the resource is marked to be processed "by value":
+        - The constructor downloads the resource (if not already available as a `localBlob`).
+        - It is then stored in the component version as a `localBlob`.
+        - The local blob is uploaded using the OCM repository's capabilities.
+
+4. **Upload Component Version**  
+   Once all resources, sources, and metadata are processed, the constructor uploads the final component version to the OCM repository.
+
+### Contract
+
+From this set of requirements we can derive the following interface for input methods:
+
+```go
+// ResourceInputMethodResult is the return value of a ResourceInputMethod.
+// It MUST contain one of
+// - ProcessedResource: The processed resource with the access type set to the resulting access type
+// - ProcessedBlobData: The local blob data that is expected to be uploaded when uploading the component version
+//
+// If the input method does not support the given input specification it MAY reject the request,
+// but if a ResourceMethodResult is returned, it MUST at least one of the above.
+//
+// If the ResourceInputMethodResult.ProcessedResource is set, the access type of the resource MUST be set to the resulting access type
+// If the ResourceInputMethodResult.ProcessedBlobData is set, the access type of the blob must be uploaded as a local resource
+// with the relation `local`, the media type derived from blob.MediaTypeAware, and the resource version defaulted
+// to the component version.
+type ResourceInputMethodResult struct {
+    ProcessedResource *descriptor.Resource
+    ProcessedBlobData blob.ReadOnlyBlob
+}
+
+// ResourceInputMethod is the interface for processing a resource with an input method declared as per
+// [spec.Resource.Input]. Note that spec.Resource's who have their access predefined, are never processed
+// with a ResourceInputMethod, but are directly added to the component version repository.
+// any spec.Resource passed MUST have its [spec.Resource.Input] field set.
+// If the input method does not support the given input specification it MAY reject the request
+//
+// The method will get called with the raw specification specified in the constructor and is expected
+// to return a ResourceMethodResult or an error.
+//
+// A method can be supplied with credentials from any credentials system by requesting a consumer identity with
+// ResourceInputMethod.GetCredentialConsumerIdentity. The resulting identity MAY be used to uniquely identify the consuming
+// method and to request credentials from the credentials system. The credentials system is not part of this interface and
+// is expected to be supplied by the caller of the input method.
+//
+// The resolved credentials MAY be passed to the input method via the credentials map, but a method MAY
+// work without credentials as well.
```

Comment:
Hmmm. I would argue that if the credentials are request then they MUST be passed in, right?

Reply:
a plugin cannot request them, it can only be provided with them. So there is no way for a plugin to say "i need credentials now" with this call. Instead they can just return an error in GetConsumerIdentity if they dont want to resolve credentials

Reply:
Ah I see. I probably misunderstood the sentence then.